nsnotifyd.1
.Dd July 10, 2015
.Dt NSNOTIFYD 1 "DNS Commands Manual"
.Os DNS
.Sh NAME
.Nm nsnotifyd
.Nd handle DNS NOTIFY messages by running a command
.Sh SYNOPSIS
.Nm
.Op Fl 46dSVw
.Op Fl l Ar facility
.Op Fl P Ar pidfile
.Op Fl u Ar user
.Op Fl R Ar min : Ns Ar max
.Op Fl r Ar min : Ns Ar max
.Op Fl s Ar authority
.Op Fl a Ar addr
.Op Fl p Ar port
.Aq Ar command
.Ao Ar zone Ac Ns ...
.Sh DESCRIPTION
The
.Nm
daemon
monitors a set of DNS
.Ar zone Ns s
and runs a
.Ar command
when any of them change.
It listens for DNS NOTIFY messages
so it can respond to changes promptly.
It also uses each zone's SOA refresh and retry parameters
to poll for updates if
.Nm
does not receive NOTIFY messages more frequently.
.Pp
You should specify zone names without the trailing dot.
The root zone can be specified as
.Ql \&.
or
.Ql root .
.Pp
Note:
.Nm nsnotify
(without
.Ql d )
is a client for sending DNS NOTIFY messages
whereas
.Nm nsnotifyd
(with
.Ql d )
is a daemon for handling DNS NOTIFY messages.
.Sh OPTIONS
.Bl -tag -width indent
.It Fl 4
Use IPv4 only
(apart from the system resolver).
.It Fl 6
Use IPv6 only
(apart from the system resolver).
.It Fl a Ar address
Listen on
.Ar address
for NOTIFY messages.
The default is
.Li 127.0.0.1 .
.Pp
You can specify an IP address or hostname.
A hostname is looked up using the system resolver.
If it resolves to multiple addresses then one
arbitrary address is chosen,
constrained by the
.Fl 4
or
.Fl 6
options.
.It Fl d
Debugging mode.
.Pp
Use once to prevent
.Nm
from daemonizing
and to make it print log messages to stderr.
.Pp
Use twice to get dumps of DNS packets.
.It Fl l Ar facility
Set the
.Xr syslog 3
facility.
The default is
.Sy daemon .
.It Fl P Ar path
Write the
.Nm
PID to the given
.Ar path
after daemonizing
and before dropping privilege.
.It Fl p Ar port
Listen on
.Ar port ,
which may be a service name or a UDP port number.
The default is the
.Sy domain
service, UDP port 53.
.It Fl R Ar interval
Override SOA
.Sy refresh
interval.
.It Fl R Ar min : Ns Ar max
Restrict SOA
.Sy refresh
intervals
to be between
.Ar min
and
.Ar max .
.It Fl r Ar interval
Override SOA
.Sy retry
interval.
.It Fl r Ar min : Ns Ar max
Restrict SOA
.Sy retry
intervals
to be between
.Ar min
and
.Ar max .
.It Fl S
Disable the SOA serial checks
.It Fl s Ar authority
Specify an authoritative server to
use for zone SOA refresh queries.
By default
.Nm
does periodic refreshes
using the system recursive resolver,
so its refresh queries may get stale cached answers.
.Pp
You can specify an IP address or hostname.
A hostname is looked up using the system resolver,
constrained by the
.Fl 4
or
.Fl 6
options.
.It Fl u Ar user
Drop privilege to
.Ar user
after daemonizing.
.It Fl V
Print details about this version of
.Nm .
.It Fl w
Accept NOTIFY messages for unknown zones
that are not given on the command line.
(Wildcard mode.)
.El
.Ss Interval syntax
Time parameters for the
.Fl R
and
.Fl r
options are in seconds,
or you can use a combination of the following time units,
as in DNS master files.
For example,
.Li 1h1m1s
is 3661 seconds.
.Pp
.Bl -tag -compact -width indent -offset indent
.It w
weeks
.It d
days
.It h
hours
.It m
minutes
.It s
seconds
.El
.Sh DETAILS
.Ss Startup
Before daemonizing,
.Nm
makes SOA queries for each
.Ar zone
to initialize its refresh and retry timers.
.Pp
Daemonizing is configured using the
.Fl P
.Ar pidfile
and
.Fl u
.Ar user
options,
or disabled with the
.Fl d
debugging option.
.Pp
When daemonizing,
.Nm
does
.Em not
change its working directory.
This allows the
.Ar command
to be context-sensitive.
.Ss Server
The
.Nm
daemon acts as a very simple UDP-only DNS server.
The only DNS queries handled by
.Nm
are NOTIFY messages.
It rejects other queries with a
.Sy REFUSED
response code, or
.Sy FORMERR
if the query is too mangled.
.Pp
Normally
.Nm
only accepts NOTIFY messages for
.Ar zones
given on the command line.
NOTIFY messages are accepted for unknown zones if you use the
.Fl w
wildcard option.
.Pp
Messages are logged via
.Xr syslog 3 .
.Ss Zone refresh
When
.Nm
receives a NOTIFY,
or when a refresh or retry timer expires,
it makes a SOA query to see if the zone has changed.
The SOA query is sent to the source of the NOTIFY
or, if a timer expired, to the server given in the
.Fl s
option.
.Pp
If the NOTIFY message was accepted for an unknown zone
because you used the
.Fl w
wildcard option,
.Nm
makes a SOA query to verify the zone exists
and to get its serial number,
and runs the command if it succeeds.
(It is unable to verify the zone has changed in this case.)
.Pp
Some jitter is applied to SOA refresh and retry timers,
so polling can occur up to 10% earlier than specified.
.Ss Command invocation
When the SOA reply indicates the zone's serial number has increased,
.Nm
runs the
.Ar command
with two or three arguments:
.Bl -enum
.It
the
.Ar zone
name without the trailing dot,
except for the root zone
.Ql \&. ;
.It
its new serial number;
.It
the source address of the NOTIFY,
or no third argument if the update was found via a periodic refresh or retry.
.El
.Pp
When the command exits successfully,
.Nm
updates its copy of the zone's SOA parameters.
It will next poll the zone on its refresh interval.
.Pp
If the SOA query or command fails,
.Nm
does not update its SOA parameters,
and and will next poll the zone on its retry interval.
.Pp
Unknown zones that were not mentioned on the command line
are not polled.
.Sh EXAMPLE - metazones
Metazones allow you to use standard DNS mechanisms -
AXFR, IXFR, NOTIFY, UPDATE -
to control the configuration of multiple name servers,
instead of using a separate out-of-band distribution system.
.Pp
For details,
see the
.Xr metazone 1
manual.
.Sh EXAMPLE - zone revision history
Say you have a zone,
.Sy example.org ,
which is updated dynamically,
and you want to automatically record its history in a
.Xr git 1
repository.
.Ss Setup git
On a server that is authoritative for
.Sy example.org ,
run the following commands:
.Bd -literal -offset indent
$ mkdir zone-history
$ cd zone-history
$ git init
$ touch example.org
$ git add example.org
$ git commit -m 'add example.org (empty)'
.Ed
.Ss Monitor the zone
The
.Nm nsnotify2git
script is designed to work with
.Nm
to record the history of a set of zones.
Continuing the transcript,
.Bd -literal -offset indent
$ nsnotifyd -P nsnotifyd.pid -p 5309 nsnotify2git example.org
.Ed
.Ss Send notifies
To configure BIND to send notifies to
.Nm ,
so it detects changes more efficiently,
look in your
.Xr named.conf 5
file for
.Bd -literal -offset indent
zone example.org {
    ...
};
.Ed
.Pp
Inside the zone clause,
add or modify the
.Ql also-notify
setting so it includes the address and port used by
.Nm ,
like
.Bd -literal -offset indent
also-notify { 127.0.0.1 port 5309; };
.Ed
.Ss Update the zone
Now, when the zone changes,
.Nm
will quickly record the change in your
.Xr git
repository.
.Bd -literal -offset indent
$ nsupdate -l
> add example.com 3600 IN TXT "foo"
> send
> quit
$ git log --format=%s
example.org IN SOA 1234
add example.org (empty)
.Ed
.Sh EXAMPLE - stealth secondary synchronization
A stealth secondary is a server which transfers authoritative copies
of a zone, but which is not listed in the zone's NS records. It will
not normally get NOTIFY messages to tell it when to update the zone,
so must rely on the zone's SOA timers instead.
.Pp
We would like stealth secondaries to get updates promptly,
but without extra manual configuration of
.Ql also-notify
lists.
.Pp
To do this,
.Nm
includes
.Nm nsnotify-liststealth
which analyzes a BIND log file to extract lists of AXFR and IXFR
clients for each zone
(excluding clients that use TSIG),
and
.Nm nsnotify
which takes zone and a list of clients that should be notified.
The
.Nm nsnotify2stealth
script bridges between
.Nm
and these two helpers.
.Ss Create working directory
The working directory contains the client lists,
one per zone,
and a symlink to the log file used by BIND.
You only need to run this command once
when creating the directory.
.Bd -literal -offset indent
$ mkdir notify-stealth
$ cd notify-stealth
$ ln -s /var/log/messages .log
.Ed
.Pp
This directory will also contain a
.Pa .pid
file for
.Nm ,
and occasionally a
.Pa .once
file to stop
.Nm nsnotify2stealth
from running more than one
.Nm nsnotify-liststealth
at a time.
.Ss Pre-populate the directory
This gets us a file per zone,
each containing a list of clients for that zone.
The
.Nm nsnotify2stealth
script will automatically update the client lists
once per day.
.Bd -literal -offset indent
$ nsnotify-liststealth .log
.Ed
.Ss Monitor the zones
Because we have a file per zone,
we can invoke
.Nm
with a glob instead of listing the zones explicitly.
The special files
.Pa ( .log .once .pid )
are dotted so that the glob works as expected.
.Bd -literal -offset indent
$ nsnotifyd -P .pid -p 5307 nsnotify2stealth *
.Ed
.Ss Send notifies
You will also need to reconfigure BIND to send notifies to
.Nm ,
as described in the previous example.
.Ss Tune BIND
If you have a lot of stealth secondaries,
.Nm nsnotify2stealth
can cause a large flood of zone transfers.
You may need to change BIND's capacity settings
as described in the ISC Knowledge Base article
cited in the
.Sx SEE ALSO
section below.
.Sh EXAMPLE - bump-in-the-wire DNSSEC
The
.Xr nsdiff 1
utility creates an
.Xr nsupdate 1
script from the differences between two versions of a zone.
It can be used as an alternative to BIND's
.Cd inline-signing
option, amongst other things.
.Pp
You can use
.Nm
together with
.Nm nsdiff
to implement a zone signer that operates as a "bump in the wire"
between a DNSSEC-unaware hidden master server and the zone's public
name servers.
.Pp
Configure your hidden master server to send notifies and allow zone
transfers to your signing server:
.Bd -literal -offset indent
also-notify { signer port 5305; };
allow-transfer { signer; };
.Ed
.Pp
Configure the signer with dynamic signed master zones,
and generate keys for them:
.Bd -literal -offset indent
zone example.org {
    type master;
    update-policy local;
    auto-dnssec maintain;
};
.Ed
.Pp
.Bd -literal -offset indent
$ dnssec-keygen -fk example.org
$ dnssec-keygen example.org
.Ed
.Pp
Run
.Nm
on the signer to trigger an update of the signed zone
as soon as an update occurs on the hidden master:
.Bd -literal -offset indent
$ nsnotifyd -P nsnotifyd.pid -p 5305 nsnotify2update example.org
.Ed
.Pp
Configure your public name servers to transfer your zones from the
signer instead of from the hidden master.
.Sh BUGS
The
.Nm
daemon is not very secure.
.Pp
It accepts any well-formed NOTIFY message,
regardless of the source.
It does not support TSIG authentication (RFC 2845)
for access control.
.Pp
The
.Nm
daemon
only handles one query at a time,
which prevents it from becoming a fork bomb.
However, you can easily overwhelm it
with more notifications than it can handle.
A spoofed NOTIFY will make
.Nm
send a SOA query to the spoofed source address
and wait for a reply (which will probably not arrive),
during which time it is unresponsive.
.Pp
You should configure
.Nm

nsnotifyd.c
/* nsnotifyd: handle DNS NOTIFY messages by running a command
 *
 * Written by Tony Finch <dot@dotat.at> <fanf2@cam.ac.uk>
 * at Cambridge University Information Services.
 *
 * You may do anything with this. It has no warranty.
 * <http://creativecommons.org/publicdomain/zero/1.0/>
 */

#define _BSD_SOURCE
#define _XOPEN_SOURCE
#define BIND_8_COMPAT
#define SYSLOG_NAMES

#include <errno.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <netinet/in.h>

#include <arpa/nameser.h>
#include <arpa/inet.h>
#include <err.h>
#include <grp.h>
#include <libgen.h>
#include <netdb.h>
#include <pwd.h>
#include <resolv.h>
#include <signal.h>
#include <syslog.h>
#include <time.h>
#include <unistd.h>

#define	log_emerg(...)   syslog(LOG_EMERG,   __VA_ARGS__)
#define	log_alert(...)   syslog(LOG_ALERT,   __VA_ARGS__)
#define	log_crit(...)    syslog(LOG_CRIT,    __VA_ARGS__)
#define	log_err(...)     syslog(LOG_ERR,     __VA_ARGS__)
#define	log_warning(...) syslog(LOG_WARNING, __VA_ARGS__)
#define	log_notice(...)  syslog(LOG_NOTICE,  __VA_ARGS__)
#define	log_info(...)    syslog(LOG_INFO,    __VA_ARGS__)
#define	log_debug(...)   syslog(LOG_DEBUG,   __VA_ARGS__)

#pragma GCC diagnostic ignored "-Wunknown-pragmas"

/* They should have used sockaddr_storage... */
typedef union res_sockaddr_union res_sockaddr_t;

typedef unsigned char byte;

#include "version.h"

static const char what_ident[] =
    "@(#) $Program: nsnotifyd $\n"
    "@(#) $Version: " VERSION " $\n"
    "@(#) $Date:    " REVDATE " $\n"
    "@(#) $Author:  Tony Finch (dot@dotat.at) (fanf2@cam.ac.uk) $\n"
    "@(#) $URL:     http://dotat.at/prog/nsnotifyd/ $\n"
;

static void
version(void) {
	const char *p = what_ident;
	for(;;) {
		while(*++p != '$')
			if(*p == '\0')
				exit(0);
		while(*++p != '$')
			putchar(*p);
		putchar('\n');
	}
}

#ifdef __GNUC__
#define dummy dummy __attribute__((unused))
#endif

static bool quit;
static bool soa_check = true;


static void
sigexit(int dummy) {
	quit = true;
}

static void
signoop(int dummy) {
}

static void
sigactions(void) {
	struct sigaction sa;
	int r;

	memset(&sa, 0, sizeof(sa));
	sa.sa_handler = SIG_IGN;
	sa.sa_flags = SA_RESTART;
	r = sigaction(SIGPIPE, &sa, NULL);
	if(r < 0) err(1, "sigaction(SIGPIPE)");
	sa.sa_handler = sigexit;
	sa.sa_flags = 0;
	r = sigaction(SIGINT, &sa, NULL);
	if(r < 0) err(1, "sigaction(SIGINT)");
	r = sigaction(SIGTERM, &sa, NULL);
	if(r < 0) err(1, "sigaction(SIGTERM)");
	sa.sa_handler = signoop;
	sa.sa_flags = 0;
	r = sigaction(SIGALRM, &sa, NULL);
	if(r < 0) err(1, "sigaction(SIGALRM)");
}

static const char *
isotime(time_t t) {
	static char buf[] = "YYYY-MM-DD HH:MM:SS +ZZZZ";
	strftime(buf, sizeof(buf), "%F %T %z", localtime(&t));
	return(buf);
}

static void
hostservstr(struct sockaddr *sa, socklen_t sa_len,
    char **host_r, char**serv_r) {
	static char host[NI_MAXHOST], serv[NI_MAXSERV];
	int r = getnameinfo(sa, sa_len,
			    host, sizeof(host), serv, sizeof(serv),
			    NI_NUMERICHOST | NI_NUMERICSERV);
	if(r) errx(1, "getnameinfo: %s", gai_strerror(r));
	*host_r = host; *serv_r = serv;
}
static char *
addrstr(struct sockaddr *sa, socklen_t sa_len) {
	char *host, *serv;
	hostservstr(sa, sa_len, &host, &serv);
	return(host);
}
static const char *
sockstr(struct sockaddr *sa, socklen_t sa_len) {
	static char hostserv[NI_MAXHOST + NI_MAXSERV];
	char *host, *serv;
	hostservstr(sa, sa_len, &host, &serv);
	snprintf(hostserv, sizeof(hostserv), "%s/%s", host, serv);
	return(hostserv);
}
static const char *
ai_sockstr(struct addrinfo *ai) {
	return(sockstr(ai->ai_addr, ai->ai_addrlen));
}

static int
listen_udp(int family, const char *addr, const char *port) {
	struct addrinfo hints, *ai0, *ai;
	int r, s;

	memset(&hints, 0, sizeof(hints));
	hints.ai_family = family;
	hints.ai_flags = AI_PASSIVE;
	hints.ai_socktype = SOCK_DGRAM;
	r = getaddrinfo(addr, port, &hints, &ai0);
	if(r) errx(1, "%s/%s: %s", addr, port, gai_strerror(r));

	for(ai = ai0; ai != NULL; ai = ai->ai_next) {
		s = socket(ai->ai_family, ai->ai_socktype, ai->ai_protocol);
		if(s < 0) {
			warn("socket %s", ai_sockstr(ai));
			continue;
		}
		r = 1;
		if(setsockopt(s, SOL_SOCKET, SO_REUSEADDR, &r, sizeof(r)) < 0) {
			warn("setsockopt %s SO_REUSEADDR", ai_sockstr(ai));
			goto next;
		}
		if(bind(s, ai->ai_addr, ai->ai_addrlen) < 0) {
			warn("bind %s", ai_sockstr(ai));
			goto next;
		}
		log_notice("listening on %s", ai_sockstr(ai));
		freeaddrinfo(ai0);
		return(s);
	next:	close(s);
	}
	errx(1, "could not listen on %s/%s", addr, port);
}

static void
res_server_name(int family, const char *name) {
	struct addrinfo hints, *ai0, *ai;

	memset(&hints, 0, sizeof(hints));
	hints.ai_family = family;
	hints.ai_flags = AI_ADDRCONFIG;
	hints.ai_socktype = SOCK_DGRAM;
	int r = getaddrinfo(name, "domain", &hints, &ai0);
	if(r) errx(1, "%s: %s", name, gai_strerror(r));
	if(ai0 == NULL) errx(1, "%s not found", name);

	int n;
	for(n = 0, ai = ai0; ai != NULL; ai = ai->ai_next, n++)
		;
	res_sockaddr_t addr[n];

	for(n = 0, ai = ai0; ai != NULL; ai = ai->ai_next, n++) {
		memset(&addr[n], 0, sizeof(addr[n]));
		memcpy(&addr[n], ai->ai_addr, ai->ai_addrlen);
	}
	res_setservers(&_res, addr, n);
	freeaddrinfo(ai0);
}

static res_sockaddr_t *res_saved_servers;
static int res_saved_server_count;

static void
res_saveservers(void) {
	int n = res_getservers(&_res, NULL, 0);
	res_saved_servers = calloc(n, sizeof(res_sockaddr_t));
	if(res_saved_servers == NULL) err(1, "malloc");
	res_saved_server_count = res_getservers(&_res, res_saved_servers, n);
}

static void
res_resetservers(void) {
	res_setservers(&_res, res_saved_servers, res_saved_server_count);
}

/*
 * When doing a timed refresh: Switch to making recursive queries via
 * the default resolver. (We always need to do this so we can look up
 * the authoritative server by name.) Make non-recursive SOA queries
 * if an authoritative server was specified on the command line.
 */
static void
soa_server_name(int family, const char *name) {
	res_resetservers();
	_res.options |= RES_RECURSE;
	if(name != NULL) {
		res_server_name(family, name);
		_res.options &= ~RES_RECURSE;
	}
}

/*
 * In response to a notify: Make a non-recursive query using the
 * server that notified us. RFC 1996 paragraph 3.11.
 */
static void
soa_server_addr(struct sockaddr *sa, socklen_t sa_len) {
	res_sockaddr_t addr;
	memset(&addr, 0, sizeof(addr));
	memcpy(&addr, sa, sa_len);
	addr.sin.sin_port = htons(53);
	res_setservers(&_res, &addr, 1);
	_res.options &= ~RES_RECURSE;
}

/*
 * Sanity checking for SOA timing parameters.
 */
uint32_t refresh_min = 1<<9;
uint32_t refresh_max = 1<<15;
uint32_t retry_min   = 1<<6;
uint32_t retry_max   = 1<<12;

static void
ttl_pair(char *str, uint32_t *min, uint32_t *max) {
	u_long ttl;
	char *sep = strchr(str, ':');
	if(sep != NULL)
		*sep++ = '\0';
	if(ns_parse_ttl(str, &ttl) < 0)
		errx(1, "invalid%s refresh time: %s",
		    sep != NULL ? " minimum" : "", str);
	*min = ttl;
	if(sep == NULL) {
		*max = *min;
		return;
	}
	if(ns_parse_ttl(sep, &ttl) < 0)
		errx(1, "invalid maximum refresh time: %s", sep);
	*max = ttl;
	return;
}

typedef struct zone {
	const char *name;
	uint32_t serial, retry;
	time_t refresh;
} zone;

static void
refresh_alarm(zone z[]) {
	zone *n;
	for(n = z; z->name != NULL; z++)
		if(n->refresh > z->refresh)
			n = z;
	if(n->name == NULL) return;
	log_debug("%s refresh at %s", n->name, isotime(n->refresh));
	alarm(n->refresh - time(NULL));
}

static void
refresh_jitter(zone *z, uint32_t interval) {
	interval -= res_randomid() % (interval / 10);
	z->refresh = time(NULL) + interval;
}

static const char *
zone_soa(zone *z) {
	byte msg[NS_PACKETSZ];
	char name[NS_MAXDNAME];
	int len, r;

	len = res_query(z->name, ns_c_in, ns_t_soa, msg, sizeof(msg));
	if(len < 0) return(hstrerror(h_errno));
	// resolver has already sanity-checked the query section
	byte *eom = msg + len, *p = msg + sizeof(HEADER);
	r = dn_skipname(p, eom);
	p += r + 4; // qname qtype qclass
	HEADER *h = (void *) msg;
	for(int ancount = ntohs(h->ancount); ancount > 0; ancount--) {
		if(p >= eom) return("truncated reply");
		p += r = ns_name_uncompress(msg, eom, p, name, sizeof(name));
		if(r < 0) return("bad owner");
		if(eom - p < 10) return("truncated RR");
		uint32_t type, class, ttl, rdlength;
		NS_GET16(type, p);
		NS_GET16(class, p);
		NS_GET32(ttl, p);
		NS_GET16(rdlength, p);
		(void)ttl;
		if(eom - p < rdlength) return("truncated RDATA");
		byte *eor = p + rdlength;
		if(strcmp(name, z->name) == 0 &&
		    class == ns_c_in && type == ns_t_soa) {
			p += r = dn_skipname(p, eor);
			if(r < 0) return("bad SOA MNAME");
			p += r = dn_skipname(p, eor);
			if(r < 0) return("bad SOA RNAME");
			if(eor - p < 12) return("truncated SOA timers");
			uint32_t refresh, retry;
			NS_GET32(z->serial, p);
			NS_GET32(refresh, p);
			NS_GET32(retry, p);
			if(refresh < refresh_min) refresh = refresh_min;
			if(refresh > refresh_max) refresh = refresh_max;
			if(retry < retry_min) retry = retry_min;
			if(retry > retry_max) retry = retry_max;
			refresh_jitter(z, refresh);
			z->retry = retry;
			return(NULL);
		}
		p = eor;
	}
	return("missing answer");
}

/* RFC 1982 */
static bool
serial_lt(uint32_t s1, uint32_t s2) {
	int64_t i1 = s1, i2 = s2, smax = 0x80000000;
	return(s1 != s2 && (
		(i1 < i2 && i2 - i1 < smax) ||
		(i1 > i2 && i1 - i2 > smax) ));
}

static void
zone_retry(zone *z) {
	if(z->retry == 0) return;
	refresh_jitter(z, z->retry);
	log_debug("%s retry at %s", z->name, isotime(z->refresh));
}

static void
zone_refresh(zone *zp, const char *cmd, const char *master) {
	zone z = *zp; // only update *zp if the refresh succeeds
	if( soa_check ) {
		const char *e = zone_soa(&z);
		if(e != NULL) {
			log_err("%s IN SOA ? %s", z.name, e);
			zone_retry(zp);
			return;
		}
		if(zp->serial == 0 && zp->refresh == 0 && zp->retry == 0) {
			log_info("%s IN SOA %d wildcard; running %s",
			    z.name, z.serial, cmd);
		} else if (serial_lt(zp->serial, z.serial)) {
			log_info("%s IN SOA %d updated; running %s",
			    z.name, z.serial, cmd);
		} else {
			log_info("%s IN SOA %d unchanged", z.name, z.serial);
			*zp = z; // refresh later
			return;
		}
	} else {
		log_info("%s running %s",
			    z.name, cmd);
	}
	switch(fork()) {
	case(-1):
		log_err("fork: %m");
		zone_retry(zp);
		return;
	case(0):;
		char serial_buf[] = "4294967295";
		snprintf(serial_buf, sizeof(serial_buf), "%u", z.serial);
		const char *cmdv[] = {
			cmd,
			z.name,
			serial_buf,
			master,
			NULL
		};
		execvp(cmd, (char**)cmdv);
		err(1, "exec %s", cmd);
	default:;
		int r;
		if(wait(&r) < 0)
			log_err("wait: %m");
		else if(!WIFEXITED(r))
			log_err("%s died with signal %d",
			    cmd, WTERMSIG(r));
		else if(WEXITSTATUS(r) != 0)
			log_err("%s exited with status %d",
			    cmd, WEXITSTATUS(r));
		else {
			*zp = z; // success
			return;
		}
		zone_retry(zp); // command failed
		return;
	}
}

static void
usage(void) {
	fprintf(stderr,
"usage: nsnotifyd [-46dV] [-l facility] [-P pidfile] [-u user]\n"
"		[-s addr] [-a addr] [-p port] command zone...\n"
"	-4		listen on IPv4 only\n"
"	-6		listen on IPv6 only\n"
"	-a addr		listen on this IP address or host name\n"
"			(default 127.0.0.1)\n"
"	-d		debugging mode\n"
"			(use twice to print DNS messages)\n"
"	-l facility	syslog facility name\n"
"	-P pidfile	write daemon pid to this file\n"
"	-p port		listen on this port number or service name\n"
"			(default 53)\n"
"	-S		disable the SOA serial checks\n"
"	-R min:max	limit SOA refresh times (default %d:%d)\n"
"	-r min:max	limit SOA retry times (default %d:%d)\n"
"	-s addr		authoritative server for refresh queries\n"
"	-u user		drop privileges to user\n"
"	-V		print version information\n"
"	-w		wildcard: accept notifies on any zone\n"
"	command		the command to run when a zone changes\n"
"	zone...		list of zones for which to accept notifies\n",
	    refresh_min, refresh_max,
	    retry_min, retry_max);
	exit(1);
}

int
main(int argc, char *argv[]) {
	int r, i;
	int family = PF_UNSPEC;
	int facility = LOG_DAEMON;
	const char *pidfile = NULL;
	const char *user = NULL;
	const char *addr = "127.0.0.1";
	const char *port = "domain";
	const char *authority = NULL;
	bool wild = false;
	char *cmd = NULL;
	int debug = 0;

	while((r = getopt(argc, argv, "46a:dl:P:p:SR:r:s:u:Vw")) != -1)
		switch(r) {
		case('4'):
			family = PF_INET;
			continue;
		case('6'):
			family = PF_INET6;
			continue;
		case('a'):
			addr = optarg;
			continue;
		case('d'):
			debug++;
			continue;
		case('l'):
			for(i = 0; facilitynames[i].c_name != NULL; i++)
				if(strcmp(facilitynames[i].c_name, optarg) == 0)
					break;
			if(facilitynames[i].c_name == NULL)
				errx(1, "%s: Unknown syslog facility", optarg);
			facility = facilitynames[i].c_val;
			continue;
		case('P'):
			pidfile = optarg;
			continue;
		case('p'):
			port = optarg;
			continue;
		case('S'):
			soa_check = false;
			continue;
		case('R'):
			ttl_pair(optarg, &refresh_min, &refresh_max);
			continue;
		case('r'):
			ttl_pair(optarg, &retry_min, &retry_max);
			continue;
		case('s'):
			authority = optarg;
			continue;
		case('u'):
			user = optarg;
			continue;
		case('w'):
			wild = true;
			continue;
		case('V'):
			version();
		default:
			usage();
		}

	openlog(basename(argv[0]), debug ? LOG_PERROR : LOG_PID, facility);

	res_init();
	res_saveservers();
	if(debug > 1) _res.options |= RES_DEBUG;
	/* be impatient */
	_res.retrans = 3;
	_res.retry = 2;

	if(debug > 1) {
		log_debug("SOA refresh limits %lu < %lu",
		    (u_long)refresh_min, (u_long)refresh_max);
		log_debug("SOA retry limits %lu < %lu",
		    (u_long)retry_min, (u_long)retry_max);
	}

	argc -= optind;
	argv += optind;
	if(argc < 1 || (argc == 1 && !wild))
		usage();

	cmd = *argv++; argc--;

	struct passwd *pw = NULL;
	if(user != NULL) {
		errno = 0;
		pw = getpwnam(user);
		if(pw == NULL && errno == 0)
			errx(1, "getpwnam %s: Unknown user", user);
		if(pw == NULL)
			err(1, "getpwnam %s", user);
	}

	soa_server_name(family, authority);
	zone zones[argc + 1];
	memset(&zones[argc], 0, sizeof(zone));
	for(zone *z = zones; argc > 0; z++) {
		memset(z, 0, sizeof(*z));
		z->name = *argv++; argc--;
		if(strcmp(z->name, "root") == 0)
			z->name = ".";
		const char *e = zone_soa(z);
		if(e != NULL) errx(1, "%s IN SOA: %s", z->name, e);
		log_info("%s IN SOA %u", z->name, z->serial);
	}

	res_resetservers();
	int s = listen_udp(family, addr, port);

	sigactions();

// don't complain about daemon() on Mac OS
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"

	if(!debug && daemon(1, 0) < 0)
		err(1, "daemon");

#pragma clang diagnostic pop

	if(pidfile != NULL) {
		FILE *fp = fopen(pidfile, "w");
		if(fp == NULL) {
			log_err("open %s: %m", pidfile);
		} else {
			fprintf(fp, "%d\n", getpid());
			fclose(fp);
		}
	}

	if(pw != NULL) {
		if(setgid(pw->pw_gid) < 0)
			log_err("setgid %s: %m", user);
		if(initgroups(pw->pw_name, pw->pw_gid) < 0)
			log_err("initgroups %s: %m", user);
		if(setuid(pw->pw_uid) < 0)
			log_err("setuid %s: %m", user);
	}

	for(;;) {
		byte msg[NS_PACKETSZ];
		char qname[NS_MAXDNAME];
		struct sockaddr_storage sa_buf;
		struct sockaddr *sa = (void *) &sa_buf;
		socklen_t sa_len;
		ssize_t len;

		refresh_alarm(zones);
		memset(msg, 0, sizeof(HEADER));
		sa_len = sizeof(sa_buf);
		len = recvfrom(s, msg, sizeof(msg), 0, sa, &sa_len);
		alarm(0);

		if(len < 0) {
			if(quit) {
				log_notice("exiting");
				if(pidfile != NULL) unlink(pidfile);
				exit(0);
			}
			if(errno != EINTR) {
				log_err("recv: %m");
				continue;
			}
			// keep refreshing until there is nothing to do
			soa_server_name(family, authority);
			bool refreshed = true;
			while(refreshed) {
				refreshed = false;
				time_t now = time(NULL);
				for(zone *z = zones; z->name != NULL; z++) {
					if(z->refresh > now)
						continue;
					log_info("%s refresh", z->name);
					zone_refresh(z, cmd, NULL);
					refreshed = true;
				}
			}
			continue;
		}
		if(debug > 1) {
			log_debug("%s query length %ld",
				  sockstr(sa, sa_len), len);
			res_pquery(&_res, msg, len, stderr);
		}
		byte *eom = msg + len;
		byte *p = msg + sizeof(HEADER);
		HEADER *h = (void *) msg;

		if(eom < p) goto formerr;
		if(h->qdcount != htons(1)) goto formerr;

		p += r = ns_name_uncompress(msg, eom, p, qname, sizeof(qname));
		if(r < 0 || eom - p < 4) goto formerr;

		int qtype, qclass;
		NS_GET16(qtype, p);
		NS_GET16(qclass, p);
		if(h->opcode != ns_o_notify ||
		    qclass != ns_c_in || qtype != ns_t_soa)
			goto refused;

		zone *z, wz = { .name = qname };
		for(z = zones; z->name != NULL; z++)
			if(strcmp(z->name, qname) == 0)
				break;
		if(z->name == NULL) {
			if(wild) z = &wz;
			else goto refused;
		}

		// TODO: call ns_verify() to check TSIG
		// however libbind's ns_verify only does HMAC MD5
		// so it needs some work...

		log_info("%s notify from %s", z->name, sockstr(sa, sa_len));
		if( soa_check )
			soa_server_addr(sa, sa_len);
		zone_refresh(z, cmd, addrstr(sa, sa_len));

		// build the reply mostly by echoing the query up to
		// p, which points to the end of the part we parsed
		h->rcode = ns_r_noerror;
	reply:
		// echo id
		h->qr = 1;
		// echo opcode
		h->aa = 1;
		h->tc = 0;
		// echo rd
		h->ra = 0;
		h->unused = 0;
		h->ad = 0;
		// echo cd
		// echo qdcount
		h->ancount = 0;
		h->nscount = 0;
		h->arcount = 0;
		// TODO: call ns_sign() to add TSIG
		if(debug > 1) {
			log_debug("%s reply length %ld",
				  sockstr(sa, sa_len), p - msg);
			res_pquery(&_res, msg, p - msg, stdout);
		}
		len = sendto(s, msg, p - msg, 0, sa, sa_len);
		if(len < 0)
			log_err("sendto %s: %m", sockstr(sa, sa_len));
		continue;
	formerr:
		log_info("%s formerr", sockstr(sa, sa_len));
		h->rcode = ns_r_formerr;
		h->qdcount = 0;
		goto reply;
	refused:
		log_info("%s refused %s %s %s", sockstr(sa, sa_len),
		       qname, p_class(qclass), p_type(qtype));
		h->rcode = ns_r_refused;
		goto reply;
	}
}
